pkg/operator/scaling.go

```diff
@@ -0,0 +1,149 @@
+	}
+}
+
+func setupScalingController(op *Operator) error {
+	objFilterOperatorConfig := namespacedNamePredicate{
+		namespace: op.opts.PublicNamespace,
+		name:      NameOperatorConfig,
+	}
+
+	err := ctrl.NewControllerManagedBy(op.manager).
+		Named("scaling").
+		WithEventFilter(predicate.ResourceVersionChangedPredicate{}).
+		For(
+			&monitoringv1.OperatorConfig{},
+			builder.WithPredicates(objFilterOperatorConfig),
+		).
+		Owns(&autoscalingv1.VerticalPodAutoscaler{}).
+		Complete(newScalingReconciler(op.manager.GetClient(), op.opts))
+	if err != nil {
+		return fmt.Errorf("scaling controller: %w", err)
+	}
+	return nil
+}
+
+func (r *scalingReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	logger, _ := logr.FromContext(ctx)
+	logger.WithValues("scaling", req.NamespacedName).Info("reconciling scaling")
+
+	var config monitoringv1.OperatorConfig
+	if err := r.client.Get(ctx, req.NamespacedName, &config); apierrors.IsNotFound(err) {
+		return reconcile.Result{}, deleteVPA(ctx, r.client, r.opts.OperatorNamespace)
+	} else if err != nil {
+		return reconcile.Result{}, fmt.Errorf("get operatorconfig: %w", err)
+	}
+
+	switch {
+	case config.Scaling.VPA.Enabled:
+		// Apply VPA
+		if err := applyVPA(ctx, r.client, r.opts.OperatorNamespace); err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{}, nil
+	default:
+		return reconcile.Result{}, deleteVPA(ctx, r.client, r.opts.OperatorNamespace)
+	}
+}
+
+func applyVPA(ctx context.Context, c client.Client, namespace string) error {
+	vpa := autoscalingv1.VerticalPodAutoscaler{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+			Name:      collectorVPAName,
+		},
+	}
+	if _, err := controllerutil.CreateOrUpdate(ctx, c, &vpa, func() error {
+		vpa.Spec = autoscalingv1.VerticalPodAutoscalerSpec{
+			TargetRef: &autoscaling.CrossVersionObjectReference{
+				APIVersion: "apps/v1",
+				Kind:       "DaemonSet",
+				Name:       collectorVPAName,
+			},
+			UpdatePolicy: &autoscalingv1.PodUpdatePolicy{
+				UpdateMode: ptr.To(autoscalingv1.UpdateModeAuto),
+			},
+			ResourcePolicy: &autoscalingv1.PodResourcePolicy{
+				ContainerPolicies: []autoscalingv1.ContainerResourcePolicy{
+					{
+						ContainerName: "prometheus",
+						Mode:          ptr.To(autoscalingv1.ContainerScalingModeAuto),
+						MinAllowed: corev1.ResourceList{
+							corev1.ResourceMemory: resource.MustParse("32Mi"),
+						},
+					},
+					{
+						ContainerName: "config-reloader",
+						Mode:          ptr.To(autoscalingv1.ContainerScalingModeOff),
+					},
+				},
+			},
+		}
+		return nil
+	}); err != nil {
+		return err
+	}
+	return nil
+}
+
+func deleteVPA(ctx context.Context, c client.Writer, namespace string) error {
+	vpa := autoscalingv1.VerticalPodAutoscaler{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      collectorVPAName,
+			Namespace: namespace,
+		},
+	}
+	if err := c.Delete(ctx, &vpa); client.IgnoreNotFound(err) != nil {
```

Comment:
nit: ditto, `return c.Delete` might be simpler

Reply:
We don't want to return a "not found" error. The logic here ignores those.